### PR TITLE
fix exprt::opX accesses in goto-instrument

### PR DESCRIPTION
**Pull request by kroening**
This improves type safety.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- n/a My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-instrument/goto_program2code.cpp [hunk 1]
```diff
@@ -110,7 +110,7 @@ void goto_program2codet::scan_for_varargs()
         to_side_effect_expr(r).get_statement() == ID_va_start)
       {
         assert(r.has_operands());
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-instrument/goto_program2code.cpp]: This assertion is now unnecessary (`to_unary_expr` already does this)

File: src/goto-instrument/dot.cpp [hunk 2]
```diff
@@ -41,7 +41,7 @@ class dott
 
   unsigned subgraphscount;
 
-  std::list<exprt> function_calls;
+  std::list<binary_exprt> function_calls;
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-instrument/dot.cpp]: Since this is specifically containing function calls, could this be a more specific type than `binary_exprt`?

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7a7609235355aa4ee277587eafbffc39b6c5bd87).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122972476

File: src/goto-instrument/goto_program2code.cpp (hunk 1 above)
Comment (kroening) [reply src/goto-instrument/goto_program2code.cpp]: removed

File: src/goto-instrument/dot.cpp (hunk 2 above)
Comment (kroening) [reply src/goto-instrument/dot.cpp]: I've made it a pair.

File: src/goto-instrument/goto_program2code.cpp [hunk 3]
```diff
@@ -315,7 +314,8 @@ goto_programt::const_targett goto_program2codet::convert_assign_varargs(
   {
     code_function_callt f(
       symbol_exprt(ID_va_start, code_typet({}, empty_typet())),
-      {this_va_list_expr, to_address_of_expr(skip_typecast(r.op0())).object()});
+      {this_va_list_expr,
+       to_address_of_expr(skip_typecast(to_unary_expr(r).op())).object()});
```
Comment (smowton) [inline src/goto-instrument/goto_program2code.cpp:318]: I'd say make an exprt for a `va_start` or do nothing, the fake unary-op isn't really adding any safety

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 494728d4b98b6eb785126264e228fa7cea5294c3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122995092

**Pull request merged**